File: src/Rstreams.c
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include <limits.h>
#include <unistd.h>
/* Only defined on Windows, etc */
#ifndef O_BINARY
#define  O_BINARY 0
#endif

#ifndef OPEN_MAX
/* Solaris' value: not set on Windows. Just needs to be large enough */
#  define OPEN_MAX 64
#endif

#include "R.h"


static char *stream_names[OPEN_MAX];
static int   stream_modes[OPEN_MAX];


static void swap(void *result, int size)
{
    int i;
    char *p = result, tmp;

    if (size == 1) return;
    for (i = 0; i < size/2; i++) {
	tmp = p[i];
	p[i] = p[size - i - 1];
	p[size - i - 1] = tmp;
    }
}


/* mode is 1 for read, 2 for write */
void openstream(char **filename, int *mode, int *handle)
{
    int fd;
    char *p = R_ExpandFileName(filename[0]);

    if(*mode == 1)
	fd = open(p, O_RDONLY|O_BINARY);
    else
	fd = open(p, O_WRONLY|O_CREAT|O_TRUNC|O_BINARY, S_IRUSR|S_IWUSR);
    if (fd == -1) error("opening `%s' failed", p);
    *handle = fd;
    stream_modes[fd] = 100 + (*mode);
    stream_names[fd] = Calloc(strlen(p)+1, char);
    strcpy(stream_names[fd], p);
}

void closestream(int *handle)
{
    int fd = *handle;
    stream_modes[fd] = 0;
    /* more important to close file than to not free memory */
    if (close(fd)) error("closing stream failed");
    Free(stream_names[fd]);
}

/* origin is 1 2 3 = "start", "current", "end" */
void seek(int *handle, int *offset, int *origin)
{
    lseek(*handle, (off_t) *offset, (*origin == 1) ? SEEK_SET :
	  (*origin == 2 ? SEEK_CUR : SEEK_END));
}


void readchar(int *handle, int *len, int *pn, char **result)
{
    int i, n = *pn;
    for (i = 0; i < n; i++)
	if (read(*handle, result[i], *len) <= 0) break;
    if (i < n) *pn = i;
}


void readint(int *handle, int *pn, int *psize, int *signd,
	     int *swapbytes, int *result)
{
    int i, n = *pn, size = *psize;
    unsigned short us1;
    short s1;
    signed char c1;
    unsigned char uc1;

    for (i = 0; i < n; i++) {
	switch (size) {
	case 4:
	    if (read(*handle, result+i, size) <= 0) goto enddata;
	    if (*swapbytes) swap(result+i, size);
	    break;
	case 2:
	    if (*signd) {
		if (read(*handle, &s1, size) <= 0) goto enddata;
		if (*swapbytes) swap(&s1, size);
		result[i] = (int) s1;
	    } else {
		if (read(*handle, &us1, size) <= 0) goto enddata;
		if (*swapbytes) swap(&us1, size);
		result[i] = (int) us1;
	    }
	    break;
	case 1:
	    if (*signd) {
		if (read(*handle, &c1, 1) <= 0) goto enddata;
		result[i] = (int) c1;
	    } else {
		if (read(*handle, &uc1, 1) <= 0) goto enddata;
		result[i] = (int) uc1;
	    }
	    break;
	default:
	    error("size must be 1, 2 or 4");
	}
    }
    return;
 enddata:
    *pn = i;
    return;
}

void readfloat(int *handle, int *pn, int *psize, int *fromint,
	       int *swapbytes, double *result)
{
    int i, n = *pn, size = *psize, *p1;
    float f1;
    double d1;
#if SIZEOF_LONG_DOUBLE > 8
    long double e1;
#endif
    unsigned int ui1, *p2;


    if (*fromint) {
	for (i = 0; i < n; i++) {
	    switch (size) {
	    case 4: /* must be unsigned */
		if (read(*handle, &ui1, size) <= 0) goto enddata;
		if (*swapbytes) swap(&ui1, size);
		result[i] = (double) f1;
		break;
	    case 8:
		if (read(*handle, &d1, size) <= 0) goto enddata;
		if (*swapbytes) swap(&d1, size);
/* I suspect these need to be swapped by endian-ness */
		if (*fromint > 1) { /* signed */
		    p1 = (int *) &d1;
		    d1 = (double)*p1 + 4294967296.0 * *(p1+1);
		} else {
		    p2 = (unsigned int *) &d1;
		    d1 = (double)*p2 + 4294967296.0 * *(p2+1);
		}
		result[i] = f1;
		break;
	    }
	}
    }
    for (i = 0; i < n; i++) {
	switch (size) {
	case 4:
	    if (read(*handle, &f1, size) <= 0) goto enddata;
	    if (*swapbytes) swap(&f1, size);
	    result[i] = (double) f1;
	    break;
	case 8:
	    if (read(*handle, &d1, size) <= 0) goto enddata;
	    if (*swapbytes) swap(&d1, size);
	    result[i] = d1;
	    break;
#if SIZEOF_LONG_DOUBLE > 8
	case sizeof(long double):
	    if (read(*handle, &e1, size) <= 0) goto enddata;
	    if (*swapbytes) swap(&e1, size);
	    result[i] = (double) e1;
	    break;
#endif
	default:
	    error("That size is unknown on this machine");
	}
    }
    return;
 enddata:
    *pn = i;
    return;
}

void readlines(int *handle, int *pn, int *bufsize, char **eol,
	       char **result)
{
    int i = 0, inpos = *bufsize, outpos	= 0, ibfs = *bufsize,
	obfs = *bufsize, match = 0, neol;
    char *inbuf, *outbuf, *eolc = *eol, null[2];

    neol = strlen(eolc);
    inbuf = Calloc(ibfs, char);
    if (neol == 0) { /* Handle \0 terminator specially */
	null[0] = 0;
	null[1] = 0;
	eolc = &(null[0]);
	neol = 1;
    }

    outbuf = Calloc(obfs, char);

    while (1) {
	if (inpos >= ibfs) { /* recharge the buffer? */
	    if ((ibfs = read(*handle, inbuf, ibfs)) <= 0) {
		/* fake an end of line marker at the end of file */
		outpos += neol - 1;
		match = neol;
		inbuf[0] = '\0';
	    }
	    inpos = 0;
	}
	if (inbuf[inpos] == eolc[match]) match++;
	else match = 0;

	if (outpos >= obfs) { /* not at eol yet, buffer not big enough */
	    obfs *= 2;
	    outbuf = Realloc(outbuf, obfs, char);
	}
	outbuf[outpos++] = inbuf[inpos++];
	if (match >= neol) {	/* matched eol string */
	    outpos -= neol;
	    result[i] = R_alloc(1, outpos + 1);
	    memcpy(result[i], outbuf, outpos);
	    result[i][outpos] = '\0';
	    if ( (ibfs > 0) | (outpos > 0) ) i++;
	    if ( (i >= *pn) | (ibfs <= 0) ) {  /* all done! */
		*pn = i;
		Free(outbuf);
		if (ibfs > 0) lseek(*handle, inpos - ibfs, SEEK_CUR);
		break;
	    }
	    outpos = 0;
	    match = 0;
	}
    }
    Free(inbuf);
}

void writechar(int *handle, int *pn, int *asciiz, char **data)
{
    int i, n=*pn;

    for (i = 0; i < n; i++)
	write(*handle, data[i], strlen(data[i]) + *asciiz);
}


void writefloat(int *handle, int *pn, int *psize, int *toint,
		int *swapbytes, double *data)
{
    int i, n = *pn, size = *psize;
    float f1;
    double d1;
#if SIZEOF_LONG_DOUBLE > 8
    long double e1;
#endif

    if (*toint) error("not yet implemented");
    /* and I don't see how to do so portably on a 32-bit machine */
    for (i = 0; i < n; i++) {
	switch (size) {
	case 4:
	    f1 = (float) data[i];
	    if (*swapbytes) swap(&f1, size);
	    write(*handle, &f1, size);
	    break;
	case 8:
	    d1 = data[i];
	    if (*swapbytes) swap(&d1, size);
	    write(*handle, &d1, size);
	    break;
#if SIZEOF_LONG_DOUBLE > 8
	case sizeof(long double):
	    e1 = (long double) data[i];
	    if (*swapbytes) swap(&d1, size);
	    write(*handle, &e1, size);
	    break;
#endif
	default:
	    error("That size is unknown on this machine");
	}
    }
}


void writeint(int *handle, int *pn, int *psize, int *swapbytes, int *data)
{
    int i, n = *pn, size = *psize;
    int i1;
    short s1;
    signed char c1;

    for (i = 0; i < n; i++) {
	switch (size) {
	case 4:
	    i1 = data[i];
	    if (*swapbytes) swap(&i1, size);
	    write(*handle, &i1, size);
	    break;
	case 2:
	    s1 = (short) data[i];
	    if (*swapbytes) swap(&s1, size);
	    write(*handle, &s1, size);
	    break;
	case 1:
	    c1 = (signed char) data[i];
	    write(*handle, &c1, size);
	    break;
	}
    }
}


/* stream info is

filename
mode
position
size

*/
void streaminfo(int *count, int *handle, char **filename, int *mode,
		int *position, int *size)
{
    int i, n = *count;
    struct stat fs;

    for (i = 0; i < n; i++) {
	mode[i] = stream_modes[handle[i]] - 100;
	if (mode[i] >= 0) {
	    strcpy(filename[i], stream_names[handle[i]]);
	    position[i] = lseek(handle[i], 0L, SEEK_CUR);
	    fstat(handle[i], &fs);
	    size[i] = fs.st_size;
	} else {
	    strcpy(filename[i], "NA");
	    position[i] = size[i] = NA_INTEGER;
	}
    }
}

void Rtruncate(int *handle, int *size)
{
    if (stream_modes[*handle] < 102)
	error("stream is not open in write mode");
    *size = lseek(*handle, 0, SEEK_CUR);
#ifdef HAVE_FTRUNCATE
    if(ftruncate(*handle, *size))
	error("file truncation failed");
#else
#  ifdef WIN32
    if(chsize(*handle, *size))
	error("file truncation failed");
#  else
    error("Unavailable on this platform");
#  endif
#endif
}

void streamcount(int *count)
{
    int i, cnt = 0;

    for (i = 0; i < OPEN_MAX; i++)
	if (stream_modes[i] > 0) cnt ++;
    *count = cnt;
}

void getstreams(int *len, int * result)
{
    int i, j=0;

    for (i = 0; i < OPEN_MAX; i++)
	if (stream_modes[i] > 0) result[j++] = i;
}

void closeallstreams()
{
    int i;

    for (i = 0; i < OPEN_MAX; i++)
	if (stream_modes[i] > 0) close(i);
}

void copystream(int *h1, int *h2, int *nbytes)
{
    char *buf;
    int nin;

    buf = Calloc(*nbytes, char);
    nin = read(*h1, buf, *nbytes);
    if (nin < 0) error("stream read error");
    write(*h2, buf, nin);
    *nbytes = nin;
    Free(buf);
}

File: R/Rstreams.R
ReadSObj <- function (s, code, len, swapbytes)
{
    if (code == 1)
        result <- as.logical(readint(s, len, swapbytes = swapbytes))
    else if (code == 2)
        result <- readint(s, len, swapbytes = swapbytes)
    else if (code == 3)
        result <- readfloat(s, len, 4, swapbytes = swapbytes)
    else if (code == 4)
        result <- readfloat(s, len, swapbytes = swapbytes)
    else if (code == 5) {
        charsize <- readint(s, swapbytes = swapbytes)
        newpos <- charsize + summary(s)$position
        result <- readchar(s, len, bufsize = charsize)
        seekstream(s, newpos)
    }
    else if (code == 6) {
        result <- as.list(integer(len))
        names(result) <- ReadSObj(s, 5, len, swapbytes)
        codes <- ReadSObj(s, 2, len, swapbytes)
        lens <- ReadSObj(s, 2, len, swapbytes)
        offsets <- ReadSObj(s, 2, len, swapbytes)
        if (all(names(result) == rep("", len)))
            names(result) <- NULL
        if (len > 0) {
            savenames <- names(result)
            names(result) <- 1:len
            for (i in 1:len) {
                seekstream(s, offsets[i])
                result <- do.call("$<-", list(result, as.character(i),
                  ReadSObj(s, codes[i], lens[i], swapbytes)))
            }
            if (length(savenames) == len)
                names(result) <- savenames
            else names(result) <- NULL
        }
    }
    else if (code == 7)
        result <- readcomplex(s, len, swapbytes = swapbytes)
    else if (code == 21) {
        temp <- ReadSObj(s, 6, len, swapbytes)
        result <- temp[[".Data"]]
        attributes(result) <- temp[-match(c(".Data", ".Dim",
            ".Dimnames", ".Label"), names(temp), nomatch = 0)]
        dim(result) <- temp[[".Dim"]]
        names(result) <- names(temp[[".Data"]])
        if (!is.null(temp[[".Label"]]))
            levels(result) <- temp[[".Label"]]
        if (!is.null(temp[[".Dimnames"]]))
            dimnames(result) <- temp[[".Dimnames"]]
    }
    else if (code == 257) {
        result <- ReadSObj(s, 5, len, swapbytes)
        mode(result) <- "name"
    }
    else if (code %in% 258:322) {
        result <- ReadSObj(s, 6, len, swapbytes)
        as.fn <- paste("as.", SModeNames[code - 256], sep = "")
        if (exists(as.fn))
            result <- eval(call(as.fn, result))
        else warning(paste("R does not support mode ", code,
            " (", SModeNames[code - 256], ")", sep = ""))
    }
    else {
        return(paste("Unrecognized S mode", code, "not supported"))
    }
    result
}

"SModeNames" <-
c("name", "string", "literal", "compiled", "(", ")", "[", "]",
"{", "}", ",", "=", "!", ":", "addop", "*/", "<dummy>", "^",
"-", "$", "logop", "&|", "<-", "->", "sp.op", " ", "repeat",
"if", "else", "break", ";", "next", "while", "for", "in", "recursive.return",
"return", "argument", "system", "end.of.file", "expression",
"system.function", "missing", "call", "function", "?", "unbalanced",
"[[", "unknown", "]]", "quit", "continue", "comment.expression",
"vector", "call(...)", "<<-", "graphics", "arg.lvalue", "internal",
"S.call", "abort", "code 318", "comment", "comment (leftover)",
"frame", "destination")

allstreams <- function ()
{
    result <- integer(.C("streamcount", count = as.integer(0),
                      PACKAGE = "Rstreams")$count)
    result <- .C("getstreams", as.integer(length(result)), result = result,
                 PACKAGE = "Rstreams")$result
    class(result) <- "stream"
    result
}

closestream <- function (stream)
{
    if (is.character(stream) && charmatch(stream, "all") == 1)
        .C("closeallstreams", PACKAGE = "Rstreams")
    else .C("closestream", as.integer(stream), PACKAGE = "Rstreams")
    invisible()
}

copystream <- function (s1, s2, nbytes)
    .C("copystream", h1 = as.integer(s1), h2 = as.integer(s2),
       nbytes = as.integer(nbytes), PACKAGE = "Rstreams")$nbytes

hexdump <- function (s, len = summary(s)$size - summary(s)$position)
{
    startpos <- summary(s)$position
    stoppos <- min(startpos + len, summary(s)$size)
    if (stoppos > startpos) {
        startdump <- 16 * (startpos%/%16)
        while (startdump < stoppos) {
            stop <- min(startdump + 16, stoppos)
            bytes <- readint(s, stop - startpos, 1, signed = FALSE)
            seekstream(s, startpos)
            chars <- readchar(s, stop - startpos, 1)
            chars[bytes < 32] <- "."
            hex <- rep("  ", 16)
            hex[(startpos - startdump + 1):(stop - startdump)] <-
                inttostr(bytes, 2)
            hex <- paste(hex, collapse = " ")
            showchars <- rep(" ", 16)
            showchars[(startpos - startdump + 1):(stop - startdump)] <- chars
            showchars <- paste(showchars, collapse = "", sep = "")
            cat(inttostr(startdump, 4), ": ", hex, " ", showchars,
                "\n", sep = "")
            startpos <- startdump + 16
            startdump <- startpos
        }
    }
    invisible()
}

inttostr <- function (x, digits = 0, base = 16, complement = FALSE)
{
    digitchars <- c(0:9, LETTERS)
    signchar <- c("-", "", "")
    x <- as.integer(x)
    signs <- sign(x)
    if (complement & base^digits/2 >= max(c(x + 1, -x))) {
        x[signs < 0] <- x[signs < 0] + base^digits
        signs <- rep(1, length(x))
    }
    else x <- abs(x)
    result <- character(length(x))
    while (digits > 0 || any(x != 0)) {
        result <- paste(digitchars[x%%base + 1], result, sep = "")
        x <- x%/%base
        digits <- digits - 1
    }
    paste(signchar[signs + 2], result, sep = "")
}

openstream <- function (filename, mode = "read")
{
    mode <- charmatch(mode, c("read", "write"))
    if (mode == 0)
        stop("Stream mode must be \"read\" or \"write\"")
    handle <- .C("openstream", as.character(filename), as.integer(mode),
        handle = integer(1), PACKAGE = "Rstreams")$handle
    if (handle < 0)
        stop(paste("Error", handle, "opening", filename))
    class(handle) <- "stream"
    handle
}

print.stream <- function (x)
{
    if (length(x) > 0) {
        info <- summary.stream(x)
        for (i in 1:length(x)) {
            if (is.na(info$position[i]))
                cat("Stream not current\n")
            else
                cat(info$filename[i], " at ", info$position[i],
                " of ", info$size[i], " bytes (",
                    c("read", "write")[info$mode[i]],
                " mode)\n", sep = "")
        }
    }
    invisible(x)
}

readchar <- function (stream, n = 1, len = NA, bufsize = 256)
{
    if (is.na(len))
        res <- .C("readasciiz", handle = as.integer(stream),
                  n = as.integer(n), as.integer(bufsize),
                  result = rep("", n), PACKAGE = "Rstreams")
    else res <- .C("readchar", handle = as.integer(stream),
                   as.integer(len), n = as.integer(n),
                   result = rep(paste(rep(" ", len), collapse = ""), n),
                   PACKAGE = "Rstreams")
    res$result[seq(len=res$n)]
}

readcomplex <- function (stream, n = 1, size = 8, swapbytes = FALSE)
{
    res <- .C("readfloat", handle = as.integer(stream),
              as.integer(2 * n), as.integer(size), as.integer(0),
              as.logical(swapbytes), result = complex(n),
              PACKAGE = "Rstreams")
    res$result[seq(len=res$n/2)]
}

readfloat <- function (stream, n = 1, size = 8, swapbytes = FALSE)
{
    res <- .C("readfloat", handle = as.integer(stream), n = as.integer(n),
       as.integer(size), as.integer(0), as.logical(swapbytes),
       result = double(n), PACKAGE = "Rstreams")
    res$result[seq(len=res$n)]
}

readint <- function (stream, n = 1, size = 4, signed = TRUE, swapbytes = FALSE)
{
    if (size < 4 | (size == 4 & signed))
        res <- .C("readint", handle = as.integer(stream),
                  n = as.integer(n), as.integer(size), as.logical(signed),
                  as.logical(swapbytes), result = integer(n),
                  PACKAGE = "Rstreams")
    else
        res <- .C("readfloat", handle = as.integer(stream),
                  n = as.integer(n), as.integer(size),
                  as.integer(1 + as.logical(signed)),
                  as.logical(swapbytes), result = double(n),
                  PACKAGE = "Rstreams")
    res$result[seq(len=res$n)]
}

readSfile <- function (filename, swapbytes = FALSE)
{
    s <- openstream(filename, "read")
    on.exit(closestream(s))
    if ((readint(s, 1, 1) == 0) && (readchar(s, 1, 6) == "S data") &&
        (readint(s, 1, 1) == 1)) {
        code <- readint(s, swapbytes = swapbytes)
        if (code < 0 | code > 65535) {
            swapbytes <- !swapbytes
            seekstream(s, -4, "current")
            code <- readint(s, swapbytes = swapbytes)
            warning(paste("Changed swapbytes to", swapbytes))
            if (code < 0 | code > 65535)
                stop("Internal error - illegal S code value\n")
        }
        len <- readint(s, swapbytes = swapbytes)
        return(ReadSObj(s, code, len, swapbytes = swapbytes))
    }
    else stop("not an S object")
}

seekstream <- function (stream, offset, origin = "start")
{
    origin <- charmatch(origin, c("start", "current", "end"))
    .C("seek", handle = as.integer(stream),
                 offset = as.integer(offset),
                 as.integer(origin), PACKAGE = "Rstreams")$offset
}

summary.stream <- function (stream)
{
    .C("streaminfo", count = as.integer(length(stream)),
       handle = as.integer(stream),
       filename = as.character(rep(paste(rep(" ", 256), collapse = ""),
       length(stream))), mode = as.integer(stream),
       position = as.integer(stream),
       size = as.integer(stream), PACKAGE = "Rstreams")
}

truncate <- function (stream)
    .C("Rtruncate", handle = as.integer(stream),
       size = as.integer(0), PACKAGE = "Rstreams")$size

writechar <- function (stream, data, asciiz = FALSE)
{
    .C("writechar", handle = as.integer(stream),
       as.integer(length(data)),
       as.integer(asciiz), as.character(data), PACKAGE = "Rstreams")
    invisible()
}

writecomplex <- function (stream, data, size = 8, swapbytes = FALSE)
{
    .C("writefloat", handle = as.integer(stream),
       as.integer(2 * length(data)), as.integer(size),
       as.integer(0), as.logical(swapbytes),
       as.complex(data), PACKAGE = "Rstreams")
    invisible()
}

writefloat <- function (stream, data, size = 8, swapbytes = FALSE)
{
    .C("writefloat", handle = as.integer(stream),
       as.integer(length(data)), as.integer(size), as.integer(0),
       as.logical(swapbytes), as.double(data), PACKAGE = "Rstreams")
    invisible()
}

writeint <- function (stream, data, size = 4, swapbytes = FALSE)
{
    if (size <= 4)
        .C("writeint", handle = as.integer(stream),
           as.integer(length(data)), as.integer(size),
           as.logical(swapbytes), as.integer(data),
           PACKAGE = "Rstreams")
    else .C("writefloat", handle = as.integer(stream),
            as.integer(length(data)), as.integer(size),
            as.integer(1), as.logical(swapbytes), as.double(data),
            PACKAGE = "Rstreams")
    invisible()
}
